integration_tests/src/main/python/hash_aggregate_test.py

```diff
@@ -605,40 +605,55 @@ def test_subquery_in_agg(adaptive, expr):
         conf = {"spark.sql.adaptive.enabled" : adaptive})
 
 
+def assert_single_level_struct(df):
+    first_level_dt = df.schema['a'].dataType
+    second_level_dt = first_level_dt['aa'].dataType
+    assert isinstance(first_level_dt, StructType)
+    assert isinstance(second_level_dt, IntegerType)
+
+# Prevent value deduplication bug across rows and struct columns
+# https://github.com/apache/spark/pull/31778 by injecting
+# extra literal columns
+#
+def workaround_dedupe_by_value(df, num_cols):
+    col_id_rng = range(0, num_cols)
+    return reduce(lambda df, i: df.withColumn(f"fake_col_{i}", f.lit(i)), col_id_rng, df)
+
+
 @allow_non_gpu(any = True)
 @pytest.mark.parametrize('key_data_gen', [
     StructGen([
-        ('a', StructGen([
-            ('aa', IntegerGen(min_val=0, max_val=9))
-        ]))], nullable=False),
+        ('aa', IntegerGen(min_val=0, max_val=9)),
+    ], nullable=False),
     StructGen([
-        ('a', StructGen([
```

Comment:
Given the issue is not specific to multi-level nesting of structs, are we just giving up trying to test multi-level struct support here?

Reply:
These tests did not intend to test deeper nesting. Per PR description a 2-level-nested struct was introduced by copy-n-paste of parameters. The same parameter generate different nesting depending on whether passed to `two_col_df` or to `gen_df` directly. The reason why it went unnoticed  is that this is a CPU fallback test without casting the struct to string first.

Reply:
Ah, sorry, I missed that in the description.  Thanks!